pageserver/src/pgdatadir_mapping.rs

```diff
@@ -1436,13 +1437,21 @@ impl<'a> DatadirModification<'a> {
     /// All the modifications in this atomic update are stamped by the specified LSN.
     ///
     pub async fn commit(&mut self, ctx: &RequestContext) -> anyhow::Result<()> {
-        let writer = self.tline.writer().await;
+        let mut writer = self.tline.writer().await;
 
         let pending_nblocks = self.pending_nblocks;
         self.pending_nblocks = 0;
 
         if !self.pending_updates.is_empty() {
-            writer.put_batch(&self.pending_updates, ctx).await?;
+            let prev_pending_updates = std::mem::take(&mut self.pending_updates);
+
+            let lsn_ordered_batch: Vec<(Key, Lsn, Value)> = prev_pending_updates
+                .into_iter()
+                .map(|(key, vals)| vals.into_iter().map(move |(lsn, val)| (key, lsn, val)))
+                .kmerge_by(|lhs, rhs| lhs.1 .0 < rhs.1 .0)
+                .collect();
+
+            writer.put_batch(lsn_ordered_batch, ctx).await?;
             self.pending_updates.clear();
```

Comment:
1. This is a behavior change: before, if `put_batch` failed, we'd still have the current batch in `pending_udpates`. Not sure that matters.
2. Would be nice to drop a quick comments aying that `put_batch` wouldl like the batch sorted. I know it's in the doc comment, but, people forget to check.

Reply:
Coment added in https://github.com/neondatabase/neon/pull/6661/commits/9a6ec4e302a03418273b81f1715f49844db3e401